client/modules/sheaf-core/src/index.ts

```diff
@@ -0,0 +1,204 @@
+import { syntaxTree, indentOnInput } from "@codemirror/language"
+import { lineNumbers } from "@codemirror/gutter"
+import { defaultKeymap, defaultTabBinding } from "@codemirror/commands"
+import { bracketMatching } from "@codemirror/matchbrackets"
+import { closeBrackets, closeBracketsKeymap } from "@codemirror/closebrackets"
+import { highlightSelectionMatches, searchKeymap } from "@codemirror/search"
+import { autocompletion, completionKeymap } from "@codemirror/autocomplete"
+import { commentKeymap } from "@codemirror/comment"
+import { rectangularSelection } from "@codemirror/rectangular-selection"
+import { redo } from "@codemirror/history"
+import { copyLineDown } from "@codemirror/commands"
+import { nextDiagnostic, openLintPanel } from "@codemirror/lint"
+
+import { writable } from "svelte/store"
+
+import { debounce } from "wj-util"
+import { printTree } from "./print-tree"
+import { confinement } from "./theme"
+import { indentHack } from "./extensions/indent-hack"
+
+interface EditorStore {
+  /** The current document of the editor. */
+  doc: EditorState["doc"]
+  /** The current 'value' (content) of the editor. */
+  value: string
+}
+
+export class SheafCore {
+  /** The element the editor is attached to. */
+  parent!: Element
+
+  /** The CodeMirror `EditorState` the editor has currently.
+   *  The state is immutable and is replaced as the editor updates. */
+  state = EditorState.create()
+
+  /** The CodeMirror `EditorView` instance the editor interacts with the DOM with. */
+  view!: EditorView
+
+  /** A store that allows reactive access to editor state. */
+  store = writable<EditorStore>({ doc: this.state.doc, value: "" })
+  subscribe = this.store.subscribe
+  set = this.store.set
+
+  /** The lines currently being interacted with by the user.
+   *  This includes all selected lines, the line the cursor is present on, etc. */
+  activeLines = writable(new Set<number>())
+
+  private spellcheckCompartment = new Compartment()
+  private guttersCompartment = new Compartment()
+
+  /** Starts the editor. */
+  async init(parent: Element, doc: string, extensions: Extension[] = []) {
+    this.parent = parent
+
+    const updateState = debounce(() => this.refresh(), 50)
+
+    const updateHandler = ViewPlugin.define(() => ({
+      update: (update: ViewUpdate) => {
+        // update store on change
+        if (update.docChanged) updateState()
+        // get active lines
+        if (update.selectionSet || update.docChanged) {
+          const activeLines: Set<number> = new Set()
+          for (const r of update.state.selection.ranges) {
+            const lnStart = update.state.doc.lineAt(r.from).number
+            const lnEnd = update.state.doc.lineAt(r.to).number
+            if (lnStart === lnEnd) activeLines.add(lnStart - 1)
+            else {
+              const diff = lnEnd - lnStart
+              for (let i = 0; i <= diff; i++) {
+                activeLines.add(lnStart + i - 1)
```

Comment:
```suggestion
              for (let lineNo = 0; lineNo <= diff; lineNo++) {
                activeLines.add(lnStart + lineNo - 1)
```
Ranges in JS has always bothered me. There should be a succinct way of doing this that doesn't involve 3rd-party libs - `activeLines.extend([lnStart..lnEnd].map(n => n -1))` or something. But alas